### DISCUSSION TITLE
Axum route can not add layer of tower

### DISCUSSION THREAD

**Discussion by Raymond0331** [tokio-rs/axum#1636] [Q&A] [answered]
create layer of tower: 

```
    let cache: Cache = Arc::new(Mutex::new(LruCache::new(1024)));
    let middleware_stack = ServiceBuilder::new()
    // timeout all requests after 10 seconds
    .timeout(Duration::from_secs(10))
    // add high level tracing of requests and responses
    .layer(TraceLayer::new_for_http())
    // compression responses
    .layer(CompressionLayer::new())
    .layer(AddExtensionLayer::new(cache))
    // convert the `ServiceBuilder` into a `tower::Layer`
    .into_inner();
```

and then axum route add this layer:

`    let app = Router::new()
    .route("/image/:spec/:url", get(generate))
    .layer(middleware_stack);
`

report an issue like the below:

error[E0277]: the trait bound `Infallible: From<Box<(dyn StdError + Send + Sync + 'static)>>` is not satisfied
   --> src/main.rs:78:12
    |
78  |     .layer(middleware_stack);
    |      ----- ^^^^^^^^^^^^^^^^ the trait `From<Box<(dyn StdError + Send + Sync + 'static)>>` is not implemented for `Infallible`
    |      |
    |      required by a bound introduced by this call
    |
    = help: the trait `From<!>` is implemented for `Infallible`
    = note: required for `Box<(dyn StdError + Send + Sync + 'static)>` to implement `Into<Infallible>`
note: required by a bound in `Router::<S, B>::layer`
   --> /Users/ray/.asdf/installs/rust/1.52.0/registry/src/github.com-1ecc6299db9ec823/axum-0.6.1/src/routing/mod.rs:319:62
    |
319 |         <L::Service as Service<Request<NewReqBody>>>::Error: Into<Infallible> + 'static,
    |                                                              ^^^^^^^^^^^^^^^^ required by this bound in `Router::<S, B>::layer`

**Comment by Raymond0331** [comment]
Sorry, posted in a wrong place. Who can help me delete it? Thanks!

**Reply by taiki-e** [reply]
(transferred discussion to tokio-rs/axum repo from tokio-rs/tokio repo.)

**Comment by davidpdrsn** [comment, marked as answer]
Are you the same person that asked in Discord?

It looks like you are using the Timeout middleware from tower, which has an Error type of [Box<dyn Error + ...>](https://docs.rs/tower/latest/tower/timeout/struct.Timeout.html#associatedtype.Error). Axum requires your middlewares to have an Error type of Infallible, because usually an error at the tower service level is not what you want (here a timeout would immediately close the client connection, instead of sending a timeout HTTP response).

You can either use axum's `HandleErrorLayer` to map the error into a response, or tower-http's timeout middleware https://docs.rs/tower-http/latest/tower_http/timeout/index.html as they don't change the service error type.
